### ISSUE TITLE
oreboot won't build for riscv64: __atomic_{load,store}_16 are not defined

### ISSUE THREAD

**Issue by rminnich** [rust-lang/rust#66240]
We have been building github.com/oreboot/oreboot with rust since last spring. Within the last month, we've hit the problem seen in https://dev.azure.com/azure0427/Oreboot%20Pipeline/_build/results?buildId=338, as well as https://github.com/oreboot/oreboot/issues/183

__atomic_store_16 is undefined, as is __atomic_load_16. 

oreboot has been working on riscv hardware (riscv64, sifive hifive board) for several months, so we're not quite sure where to start. I've been looking around for similar problems but nothing quite fits.

Thanks.

**Comment by mati865** [comment]
Already reported to upstream crate: https://github.com/rust-lang/compiler-builtins/issues/322

**Comment by rminnich** [comment]
here is the "fix" for oreboot

`+.globl __atomic_store_16
+__atomic_store_16:
+       ret
+.globl __atomic_load_16
+__atomic_load_16:
+       ret
`

i.e. fake up no op atomic load and store for 16 bits, b/c the architecture does not support them. 

You're asking the architecture to do something it can't do ... :-p

**Comment by rminnich** [comment]
sorry about the format, that's what github claims is a code copy/paste?
`
+.globl __atomic_store_16
+__atomic_store_16:
+       ret
+.globl __atomic_load_16
+__atomic_load_16:
+       ret
 .globl _stack_ptr
 .section .rodata
 _stack_ptr: .word 0x08010000
`

**Comment by asb** [comment]
Copying comment from https://github.com/oreboot/oreboot/issues/183#issuecomment-552965077
CC @lenary
In the GCC world this symbol is provided by libatomic rather than libgcc directly. In the LLVM world, compiler-rt provides the symbol. Rust should either use a RISC-V compiler-rt or if using libgcc it should link libatomic as well.

**Comment by lenary** [comment]
Yeah. This issue comes up every so often, especially more recently.

The RISC-V "A" extension only introduces instructions for 32- and 64-bit atomic operations. If you need up-to-32-bit atomic operations (ie, 8-bit or 16-bit atomic operations), then you have to link with `-latomic` if using gcc's compiler builtins (because libgcc and libatomic are separate libraries).

LLVM's compiler-rt is a slightly different, and contains both the compiler builtins and the atomic libcalls in the same library, so this issue should arise less once LLVM can switch to compiler-rt for RISC-V. That could happen soon given https://reviews.llvm.org/D68393 has landed.

**Comment by mati865** [comment]
I was fixed in compiler-builtins crate via https://github.com/rust-lang/compiler-builtins/pull/324.
Can you make xargo pull newer version? Otherwise it may require updating this dependency in Rust repo.

**Comment by istankovic** [comment]
@rminnich @lenary is this issue now fixed?

**Comment by workingjubilee** [comment]
cc @kito-cheng @michaelmaitland @robin-randhawa-sifive Is this issue still an active concern?

**Comment by taiki-e** [comment]
First, the suffix of the GCC atomic builtins is in bytes, not in bits, so `__atomic_*_16` is 16-byte (128-bit), not 16-bit.

And, there is a relatively recently ratified Zacas extension for 128-bit atomic in RISC-V, but it did not exist then and in any case LLVM generates libcalls when 128-bit atomic intrinsics are used on RISC-V.

So, I believe this has been fixed by https://github.com/rust-lang/compiler-builtins/pull/324, which fixed the problem of calling 128-bit atomic intrinsics on unsupported platforms. (I began writing RISC-V-related code later than that fix, but at least I have never seen "undefined symbol" error about atomics when compiling binaries for RISC-V bare-metal/Linux, except for a regression I fixed in https://github.com/rust-lang/rust/pull/114497.)

(Btw, AFAIK there are no ratified 16-bit atomic instructions (AMO,LR/SC) in RISC-V except those related to the recently ratified Zabha extension, but 16-bit load/store are exist, and it is possible to implement 16-bit CAS/RMW using 32-bit atomic instructions, and LLVM actually does so.)
